mitodl/ocw-studio
# Check external resources for broken links

### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/3922

### Description (What does it do?)

This PR implements the plan described in https://github.com/mitodl/hq/issues/3922.

What still needs to be done (this was my plan, you may alter it as needed):

- [x] Create a new `external_resources` app.
- [x] Define a model.
- [x] Write logic that checks if a link is broken.
- [x] Add an admin model.
- [x] Test the current implementation to see if it works and/or makes sense.
- [ ] Celery/Task implementation
  - [x] Create a celery workflow. Batch tasks, and define rate limits and task priorities.
  - [ ] Add a new celery queue dedicated to the new low-priority tasks.
  - [ ] The prod and rc infrastructure have 2 workers. Only one of them should process tasks from our new queue. With the assistance of DevOps, add the queue to one of the workers. (i.e. extra_worker).
  - [ ] Update the docker-compose definition to add the new queue to the worker. 
- [ ] Do any data-platform changes, if needed. 
- [x] Test the functionality and write unit tests.

### How can this be tested?
1. switch to branch 'hussaintaj/3922-check-external-resources'
2. Restart containers.
3. Apply DB migrations: `docker-compose run web ./manage.py migrate`
4. From celery logs, verify a repeated task should be added in the celery with name `check-broken-external-urls`. Task is set to repeat every week. Frequency can be modified by changing `CHECK_EXTERNAL_RESOURCE_STATUS_FREQUENCY` in main/settings.py
5. Go to http://localhost:8043/admin/external_resources/externalresourcestate/. This should be created after migrations applied 
6. External Resource States should have been populated with existing url checks after first execution of task

### Additional Testing
#### To manually test the functionality locally with a batch of urls
1. Spin up containers
2. Get into Django shell: `docker-compose exec web ./manage.py shell`
3. Run following commands to import necessary code: 
     - `from external_resources.tasks import check_external_resources`
     - `from websites.models import WebsiteContent`
     - `from websites.constants import CONTENT_TYPE_EXTERNAL_RESOURCE`
4. Get website content data:
    `external_resources = list(WebsiteContent.objects.filter(type=CONTENT_TYPE_EXTERNAL_RESOURCE).values_list("id", flat=True))`
5. (Optional) check length of data and splice if necessary: `len(external_resources)` and `external_resources = external_resources[:1]`
6. run checking task: `check_external_resources(external_resources)`
7. to validate resource was updated or created, retrieve resource from.DB:
    - `resources = WebsiteContent.objects.filter(id__in=external_resources).select_related("external_resource_state")`
    - `resource = next(resources.iterator())`
    - `str(resource.external_resource_state.last_checked)` 
8. repeat step 7 to update and validate time stamps again.

#### To manually execute task in celery
1. Spin up containers
2. Get into Django shell: `docker-compose exec web ./manage.py shell`
3. Run following commands to import necessary code: 
     - `from external_resources.tasks import check_external_resources_for_breakages`
     - `results = check_external_resources_for_breakages.apply()`
** This might produce large number of urls to check and may take some time to complete.
